### PR TITLE
[Bugfix][ROCm] Keep gfx942 bf16 persistent MLA decode at qlen <= 4

### PR DESCRIPTION
**Pull request by mustafayildirim**
## Purpose

Fixes the decode-path half of #55609: `ROCM_AITER_MLA` has no persistent bf16 MLA-decode kernel for `query_len > 4` on gfx942 (MI300X) — `get_heuristic_kernel_mla` fails (SIGABRT pre-ROCm/aiter#5297, `RuntimeError` after) and kills the MTP-verify batch. 16 `spec_decode_medium` cases in `tests/v1/attention/test_mla_backends.py` fail this way; `spec_decode_small` (qlen ≤ 4) and fp8 KV are unaffected.

## Root cause

`AiterMLAMetadataBuilder`'s persistent-metadata predicate treats `decode_num_heads >= 16` as sufficient for bf16 multi-token queries:

```python
and (
    not causal
    or self._decode_num_heads >= AiterMLAHelper._AITER_MIN_MLA_HEADS   # unsafe on gfx942
    or max_qo_len <= AiterMLAHelper._ASM_PADDED_MAX_PS_QLEN
    or is_quantized_kv_cache(...)
)
```

`query_len > 4` with gqa 16 reaches a persistent kernel only through aiter's q-row fold onto the **(gqa 32, qseqlen 4)** kernel. Checking the shipped manifests (`hsa/<arch>/mla/mla_asm.csv`, bf16/bf16 persistent entries):

| arch | persistent (Gqa, qSeqLen) | fold target (32, 4) shipped? |
|---|---|---|
| gfx942 | (8, 2), (16, 4) | ❌ |
| gfx950 | (16, 4), (32, 4), (64, 1), (96, 4) | ✅ |

So the heads≥16 arm is gfx950-only; on gfx942 it requests a kernel that doesn't exist. The existing comment already says "where the gfx950 fold is absent" — the guard just never checked the arch.

## Fix

Gate the heads≥16 arm on gfx950 (new `_asm_persistent_mla_heads16_qlen8_supported()` helper, same shape as the neighboring `_gluon_mla_decode_supported` / `_aiter_mla_non_causal_asm_kernels` arch probes). On gfx942, qlen>4 bf16 now builds **non-persistent** metadata, and aiter's non-persistent (gqa 16, qseqlen 8) entry — which gfx942 *does* ship — serves the batch. This is option B from the issue: never request the missing kernel. The `o.zero_()` guard for qlen>1 non-persistent verify already exists.

Unchanged: gfx950 (helper returns True), non-causal blocks, fp8/quantized KV, and qlen ≤ 4 on both arches.

## Verification

No gfx942 box available; verified on gfx950 (MI350X) that the change is behavior-preserving there, plus manifest-level differential for the gfx942 claim:

- `tests/v1/attention/test_mla_backends.py` (upstream main, patched module overlaid into the live install), `-k "backend_correctness and spec_decode and ROCM_AITER and not fp8"`: **12 passed** with the fix, 12 passed baseline (unchanged on gfx950 — the helper returns True there, as intended).
- Live: `_asm_persistent_mla_heads16_qlen8_supported()` → `True` on gfx950.
- Manifest check (aiter main, `hsa/*/mla/mla_asm.csv`): gfx942 persistent bf16 = {(8,2), (16,4)} — no qlen>4; gfx942 non-persistent (16, 8) **exists** (the fallback target).
- Guard truth table checked for all six regime combinations (gfx942/950 × causal/qlen/dtype).

Real gfx942 confirmation still needs an MI300X run — happy to hand the repro command from the issue to whoever has one: `VLLM_ROCM_USE_AITER=1 pytest -q tests/v1/attention/test_mla_backends.py -k "backend_correctness and spec_decode_medium"` (the 16 previously-failing cases should pass through the non-persistent path).

### CODE REVIEW

**Comment by mustafayildirim** [comment]
## gfx942 (MI325X) verification — and a correction to the fallback premise

Ran the full differential on MI325X (gfx942) against **both** aiter builds, using a clean full-package shadow of upstream/main+this-PR over the nightly image (no file-level mixing):

**Baseline (no fix):**

| aiter | Result |
|---|---|
| 0.1.21.post1 (vllm nightly image pin) | **16 failed — SIGABRT** kills the pytest worker at first `spec_decode_medium` case (exactly the issue's count) |
| 0.1.23.dev88 (main, installed in-pod) | **aborts** — `.FFFFFFFF` `Fatal Python error: Aborted` |

**With commit 1 (guard routes gfx942 qlen>4 bf16 to the non-persistent entry):**

| aiter | Result | kernel dispatched |
|---|---|---|
| 0.1.21.post1 | 8 pass / **8 fail** — wrong output, max diff 0.11–0.16 vs SDPA | `mla_dec_stage1_bf16_a16w16_subQ128_mqa128` (a **Gqa=128** kernel answering a Gqa=16/qseqlen-8 call) |
| main | 8 pass / **8 fail** — wrong output, same magnitude | `mla_a16w16_qh16_m32x4` |

So the abort is fixed, but the issue's assumption that "the non-persistent (gqa 16, qseqlen 8) entry covers it" does **not** hold: the manifest row exists but computes wrong verify output on gfx942 in every aiter build tested. `spec_decode_small` (qlen 2) passes 8/8 on both. aiter's own tests never cover this shape (their decode kernel test covers nhead16 only at mtp=1 and nhead128 at mtp=2; the persistent test states "supported: decode_qlen = 1,2,3,4"), and `mla.py`'s fold path requires persistent mode and skips nhead=16 entirely — so there is no working route.

**Commit 2 therefore adds a hard bound**: non-persistent causal bf16 decode with `max_qo_len > 4` now raises a `ValueError` naming the mitigations instead of emitting garbage logits.

**With both commits:**
- gfx942: 8 pass (qlen≤4), 8 fail — all failures are the clean `ValueError`, zero numerical mismatches
- gfx950 (MI350X): **48/48 pass** (`-k "backend_correctness and spec_decode and ROCM_AITER and 1.0-1.0"`) — the bound never fires, persistent path unchanged; earlier full run (96/96 incl. fp8) also clean
- qlen ≤ 4 unchanged on both arches

The proper perf fix for gfx942 qlen>4 needs a working aiter kernel (persistent fold target or a fixed non-persistent entry) — filed upstream with ROCm/aiter with the dispatch evidence above.

**Comment by mustafayildirim** [comment]
Aiter-side tracking: ROCm/aiter#5679 (wrong output from the non-persistent bf16 nhead16/qseqlen-8 path on gfx942, with the dispatch evidence from both aiter builds). The hard bound here can be relaxed once aiter serves that shape.

**Comment by mustafayildirim** [comment]
Ping for review — this fixes the gfx942 SIGABRT from #55609 that kills MTP-verify batches. Full evidence on the PR:

- **gfx942 (MI325X)**: baseline 16/16 aborts → with fix, qlen≤4 passes and qlen>4 now raises a clean `ValueError` (the previously-dispatched aiter fallback kernel returns wrong output — both aiter 0.1.21.post1 and main; tracked as ROCm/aiter#5679)
- **gfx950 (MI350X)**: 48/48 pass on the same selection, 96/96 including fp8 — no behavior change

@gdhyi @dllehr-amd — could one of you take a look when you get a chance?
